src/strong_ptr.h

```diff
@@ -0,0 +1,317 @@
+        {
+            // Some long-lived operation
+        }
+
+        int main()
+        {
+            strong_ptr<int> str(make_strong<int>(100));
+            std::thread(func, str.get_shared());
+            decay_ptr<int> dec(std::move(strong));
+            // The original strong_ptr is now reset
+            while (!dec.decayed()) {
+                // wait here
+                continue;
+            }
+            // dec is guaranteed to be unique here. Its memory will be freed
+            // when it goes out of scope.
+            return 0;
+        }
+*/
+
+template <typename T>
+class decay_ptr;
+
+template <typename T>
+class strong_ptr
+{
+    struct shared_deleter {
+        void operator()(T* /* unused */) const {}
+        const std::shared_ptr<void> m_data;
+    };
+
+    friend class decay_ptr<T>;
+
+    template <typename U, typename... Args>
+    friend strong_ptr<U> make_strong(Args&&... args);
+
+    template <typename U>
+    explicit strong_ptr(std::shared_ptr<U>&& rhs) : m_data{std::move(rhs)}
+    {
+    }
+
+public:
+    using element_type = T;
+
+    constexpr strong_ptr() : strong_ptr(nullptr) {}
+
+    constexpr strong_ptr(std::nullptr_t) : m_shared{nullptr} {}
+
+    template <typename U>
+    explicit strong_ptr(U* ptr) : m_data(ptr)
+    {
+    }
+
+    template <typename Deleter>
+    strong_ptr(std::nullptr_t ptr, Deleter deleter) : m_data{ptr, std::move(deleter)}
+    {
+    }
+
+    template <typename U, typename Deleter>
+    strong_ptr(U* ptr, Deleter deleter) : m_data{ptr, std::move(deleter)}
+    {
+    }
+
+    template <typename U, typename Deleter>
+    strong_ptr(std::unique_ptr<U, Deleter>&& rhs) : m_data{std::move(rhs)}
+    {
+    }
+
+    template <typename U>
+    strong_ptr(strong_ptr<U>&& rhs) : m_data{std::move(rhs.m_data)}, m_shared{std::move(rhs.m_shared)}
+    {
+    }
+
+    strong_ptr(strong_ptr&& rhs) noexcept : m_data{std::move(rhs.m_data)}, m_shared{std::move(rhs.m_shared)} {}
+
+    ~strong_ptr() = default;
+
+    strong_ptr(const strong_ptr& rhs) = delete;
+    strong_ptr& operator=(const strong_ptr& rhs) = delete;
+
+    strong_ptr& operator=(strong_ptr&& rhs) noexcept
+    {
+        m_data = std::move(rhs.m_data);
+        m_shared = std::move(rhs.m_shared);
+        return *this;
+    }
+
+    template <typename U>
+    strong_ptr& operator=(strong_ptr<U>&& rhs)
+    {
+        m_data = std::move(rhs.m_data);
+        m_shared = std::move(rhs.m_shared);
+        return *this;
+    }
+
+    template <typename U>
+    strong_ptr& operator=(U* rhs)
+    {
+        reset(rhs);
+        return *this;
+    }
+
+    template <typename U, typename Deleter>
+    strong_ptr& operator=(std::unique_ptr<U, Deleter>&& rhs)
+    {
+        m_data = std::move(rhs);
+        m_shared.reset(nullptr, shared_deleter{m_data});
+        return *this;
+    }
+
+    void reset()
+    {
+        m_data.reset();
+        m_shared.reset();
+    }
+    void reset(std::nullptr_t)
```

Comment:
What is the advantage of having a specific overload for nullptr_t?

Reply:
Heh, I have no idea why I added that. Will remove.